### PR TITLE
260 Refactor options

### LINKED ISSUE #260: Refactor the software options dictionary to make it more user friendly
The current method of defining options is both complicated for the user and hard to maintain.

Options should be controlled in a single place so that setting an option is consistent, with priority:
1. Values set in the example script.
2. Values set in the options file.
3. Values set as a default programmatically.

Options should also be able to be set easily, and have clearly defined default values.

### PR DESCRIPTION
**Pull request by AndrewLister-STFC**
#### Description of Work

Fixes #260 

Refactors the options and changes the input format to a .ini file.
The .ini file improves on the json version as it allows comments, and is much more human readable.

The options are now implemented as a class, and self document based on the default values.
This format allows a priority which is consistent by design rather than convention and is as follows:
1. Values set in code
2. Values set in options file
3. Values predefined in the options class (default)

#### Testing Instructions

1. Run the code with varying options and check that results are as expected
2. Read through the docs [here](https://fitbenchmarking.readthedocs.io/en/260-options_refactor/users/options.html) to check that they make sense.

Function: Does the change do what it's supposed to?

Tests: Does it pass? Is there adequate coverage for new code?

Style: Is the coding style consistent? Is anything overly confusing?

Documentation: Is there a suitable change to documentation for this change?

### CODE REVIEW

**Comment by AndrewLister-STFC** [comment]
Blocked pending the decision on #310 and if #311 is merged

**Comment by AndrewLister-STFC** [comment]
After chat with @wathen still need to move the links into the default options rather than the docs

File: docs/source/users/options.rst [hunk 1]
```diff
@@ -4,179 +4,28 @@
...
-The software to use in fitting as a string or list of strings.
-Selected softwares will be benchmarked.
+The default options are a complete set of options with sensible values.
+These will be used when no other values are given for any of the options,
+the values for these can be seen at the end of this document.
 
-Available options are ``"dfogn"``, ``"mantid"``, ``"minuit"``, ``""ralfit""``, ``"sasview"``,
-and ``"scipy"``.
+The options file must be a `.ini` file, and a good reference for this can be
+found in the examples, as well as at the bottom of this document.
```
Comment (wathen) [inline docs/source/users/options.rst]: Maybe worth adding a link to a formatting page?

File: example_scripts/example_runScripts.py [hunk 2]
```diff
@@ -12,64 +12,51 @@
...
-    # None will default to the list on the options file
-    custom_minimizers = None
-
-    # SPECIFY THE MINIMIZERS YOU WANT TO BENCHMARK, AND AS A MINIMUM FOR THE SOFTWARE YOU SPECIFIED ABOVE
+    # Find the options file
     current_path = os.path.dirname(os.path.realpath(__file__))
     if len(args) > 1:
         # Read custom minimizer options from file
-        software_options['minimizer_options'] = None
-        software_options['options_file'] = current_path + args[1]
```
Comment (wathen) [inline example_scripts/example_runScripts.py]: Shall we use argparse? Could be a different way to add options down the line too

File: docs/source/users/options.rst [hunk 3]
```diff
@@ -4,179 +4,28 @@
...
-------------
-The software to use in fitting as a string or list of strings.
-Selected softwares will be benchmarked.
+The default options are a complete set of options with sensible values.
+These will be used when no other values are given for any of the options,
+the values for these can be seen at the end of this document.
 
-Available options are ``"dfogn"``, ``"mantid"``, ``"minuit"``, ``""ralfit""``, ``"sasview"``,
-and ``"scipy"``.
+The options file must be a `.ini` file, and a good reference for this can be
```
Comment (wathen) [inline docs/source/users/options.rst]: Maybe worth adding an external link to the ini file format

File: example_scripts/example_runScripts.py (hunk 2 above)
Comment (wathen) [inline example_scripts/example_runScripts.py]: Shall we move to argparse? Would be another way in the future to add options

File: fitbenchmarking/fitbenchmark_one_problem.py [hunk 4]
```diff
@@ -14,76 +14,88 @@
 from fitbenchmarking.fitting.plotting import plot_helper, plots
```
Comment (wathen) [inline fitbenchmarking/fitbenchmark_one_problem.py]: Can't seem to add comment above but lines 7-9 are not alphabetical order

File: fitbenchmarking/results_output.py [hunk 5]
```diff
@@ -78,8 +62,8 @@ def save_results_tables(software_options, results_per_test, group_name,
                                             weighted_str,
```
Comment (wathen) [inline fitbenchmarking/results_output.py]: Could you change this to be `group_name,x, weighted_str`?

File: docs/source/users/options.rst (hunk 3 above)
Comment (wathen) [inline docs/source/users/options.rst]: Maybe worth adding an external link to ini file format?

File: example_scripts/example_runScripts.py [hunk 6]
```diff
@@ -12,64 +12,51 @@
...
-    elif custom_minimizers:
-        # Custom minimizer options:
-        software_options['minimizer_options'] = custom_minimizers
+        options_file = os.path.join(current_path, args[1])
```
Comment (wathen) [inline example_scripts/example_runScripts.py]: Shall we use argparse? Could use that to add options later as well

File: fitbenchmarking/fitbenchmark_one_problem.py (hunk 4 above)
Comment (wathen) [inline fitbenchmarking/fitbenchmark_one_problem.py]: Imports on lines 7-9 are not in alphabetical order

File: fitbenchmarking/results_output.py [hunk 7]
```diff
@@ -14,7 +14,7 @@
 
 from fitbenchmarking.utils.logging_setup import logger
 from fitbenchmarking.resproc import visual_pages
-from fitbenchmarking.utils import create_dirs, options, misc
+from fitbenchmarking.utils import create_dirs
```
Comment (wathen) [inline fitbenchmarking/results_output.py]: Should these be in alphabetical order as well?

File: fitbenchmarking/results_output.py [hunk 8]
```diff
@@ -25,51 +25,35 @@
 HTML_COLOUR_SCALE = ['#fef0d9', '#fdcc8a', '#fc8d59', '#e34a33', '#b30000']
```
Comment (wathen) [inline fitbenchmarking/results_output.py]: Can you add this to the options file?

File: fitbenchmarking/utils/tests/test_create_dirs.py [hunk 9]
```diff
@@ -12,33 +12,15 @@
```
Comment (wathen) [inline fitbenchmarking/utils/tests/test_create_dirs.py]: Imports are not in alphabetical order

File: docs/source/users/options.rst [hunk 10]
```diff
@@ -4,179 +4,28 @@
...
+----------------
+This is a template you can use which contains information on each option
+available, as well as the defaults.
 
-Number of runs defines how many times FitBenchmarking calls a minimizer and
-calculates an average elapsed time using ``timeit``.
 
-Default set as ``5``.
+.. include:: ../../../fitbenchmarking/utils/default_options.ini
+   :literal:
```
Comment (wathen) [inline docs/source/users/options.rst]: When I view this file I can see the default options

File: example_scripts/example_runScripts.py (hunk 2 above)
Comment (AndrewLister-STFC) [reply example_scripts/example_runScripts.py]: I think that should be a separate work package

File: fitbenchmarking/fitbenchmark_one_problem.py (hunk 4 above)
Comment (AndrewLister-STFC) [reply fitbenchmarking/fitbenchmark_one_problem.py]: done

File: fitbenchmarking/results_output.py (hunk 5 above)
Comment (AndrewLister-STFC) [reply fitbenchmarking/results_output.py]: done

File: fitbenchmarking/results_output.py (hunk 7 above)
Comment (AndrewLister-STFC) [reply fitbenchmarking/results_output.py]: yes

File: fitbenchmarking/utils/tests/test_create_dirs.py (hunk 9 above)
Comment (AndrewLister-STFC) [reply fitbenchmarking/utils/tests/test_create_dirs.py]: done

File: fitbenchmarking/results_output.py (hunk 8 above)
Comment (wathen) [reply fitbenchmarking/results_output.py]: Will be done separately

File: docs/source/users/options.rst (hunk 1 above)
Comment (AndrewLister-STFC) [reply docs/source/users/options.rst]: done

**Pull request merged**